python/src/aiconfig/callback.py

```diff
@@ -0,0 +1,139 @@
+# Standard Libraries
+import asyncio
+import json
+import time
+
+# Third Party Libraries
+from result import Err, Ok
+from pydantic import BaseModel, ConfigDict
+from typing import (
+    Coroutine,
+    Final,
+    List,
+    Sequence,
+    TextIO,
+    TypeAlias,
+    Union,
+    Any,
+    Awaitable,
+    Callable,
+    Optional,
+)
+
+# Constants
+DEFAULT_TIMEOUT = 5  # Default timeout for callback execution in seconds
+
+
+class CallbackEvent:
+    """
+    Represents an event with data to be passed to a callback.
+    """
+
+    def __init__(self, name: str, data: Any, ts_ns: int = time.time_ns()):
+        self.name = name
+        # Anything available at the time the event happens.
+        # It is passed to the callback.
+        self.data = data
+        self.ts_ns = ts_ns
+
+
+# Type Aliases
+Callback = Callable[[CallbackEvent], Awaitable]
+Result: TypeAlias = Union[Ok[Any], Err[Any]]
+
+
+async def execute_coroutine_with_timeout(
+    coroutine: Coroutine[Any, Any, Any], timeout: int
+) -> Result:
+    """
+    Execute a coroutine with a timeout, return an Ok result or an Err on Exception
+
+    Args:
+        coroutine: The coroutine to execute
+        timeout: The timeout in seconds
+    
+    Returns:
+        An Ok result if the coroutine executes successfully or an Err on Exception
+    """
+    try:
+        task = asyncio.create_task(coroutine)
+        res = await asyncio.wait_for(task, timeout=timeout)
+        return Ok(res)
+    except Exception as e:
+        return Err(str(e))
+
+
+class Record(BaseModel):
+    model_config = ConfigDict(strict=True, frozen=True)
+
+
+class pydanticBackedCallBackEvent(Record):
+    """
+    Data class that represents an event that triggers a callback. A copy of CallbackEvent
+    but backed by a pydantic model for easy serialization.
+    """
+    name: str
+    data: Any
+    ts_ns: int = time.time_ns()
+
+
+class CallbackManager:
+    """
+    Manages a sequence of callbacks to be executed in response to Events
+    """
+    def __init__(self, callbacks: Sequence[Callback], timeout: int = DEFAULT_TIMEOUT) -> None:
+        self.callbacks: Final[Sequence[Callback]] = callbacks
+        self.results: List[Result] = []
+        self.timeout = timeout
+
+    async def run_callbacks(self, event: CallbackEvent) -> None:
+        event = pydanticBackedCallBackEvent(**event.__dict__)
```

Comment:
Arent you going to have a lot of keys in here that will break the constructor?

Reply:
I'm not sure I follow on how it could break, all keys will be under `data` in 

```
CallbackEvent:
     data: Any
 ```
 
The reason behind the second class def is that constructing classes with Pydantic Base Models is a slight inconvenience. There isn't really a need for pydantic here besides easy JSON serialization. That way the user can just create the regular CallbackEvent class and it gets converted into the Pydantic BaseModel version here for the benefits. Will update the docstring/comment

Could this end up being a bad thing?

Reply:
I was able to just have CallbackEvent be a Record (which is a type of BaseModel) without issues. No need to convert here. Just instantiate at callsite, e.g.

```
...run_callbacks(
            CallbackEvent(
                name="openai_embeddings_fragments_embedded",
                data=dict(
                    n_batches=len(batches),
                    n_embeddings=len(out),
                ),
            )
```


are you having serialization issues?